<launch>
    <!-- This launchfile launches the full Guidance, Navigation, and Control system. -->
    <!-- This includes the cascaded pid controller, teleop node, and (TODO) the trajectory planner. -->
    <!-- TODO: Support the keyboard controller via a topic mux on the wrench output-->

    <!-- Name of the robot model  -->
    <arg name="model_name" default="albatross"/>
    <!-- path to config package  -->
    <arg name="config_pkg" default="$(eval find(arg('model_name') + '_description'))"/>
    <!-- force topic (for the thruster manager) to publish to -->
    <arg name="thruster_manager_topic" default="/$(arg model_name)/thruster_manager/input"/>
    <!-- odom topic to listen on -->
    <arg name="odom_topic" default="odom"/>

    <group>
        <!-- Launch State Estimation -->
        <node pkg="robot_localization" type="ekf_localization_node" name="ekf_se" clear_params="true" output="screen">
            <rosparam command="load" file="$(arg config_pkg)/ekf_params.yaml"/>
        </node>

        <node pkg="tauv_common" type="state_estimation" name="state_estimation" clear_params="true" ns="$(arg model_name)" output="screen"/>


        <!-- Load controller params: -->
        <rosparam file="$(arg config_pkg)/control_params/inertial.yaml" command="load"/>
        <rosparam file="$(arg config_pkg)/control_params/vel_pid_control.yaml" command="load"/>
        <rosparam file="$(arg config_pkg)/control_params/pos_pid_control.yaml" command="load"/>
        <rosparam file="$(arg config_pkg)/control_params/default_config.yaml" command="load"/>

        <!-- Launch controllers -->
        <node pkg="tauv_common" type="acceleration_controller" name="acceleration_controller"
              output="screen">
            <remap from="thruster_manager/input" to="$(arg thruster_manager_topic)"/>
        </node>
        <node pkg="tauv_common" type="velocity_controller" name="velocity_control"
              output="screen">
            <remap from="cmd_accel" to="controller/cmd_acc"/>
            <remap from="odom" to="$(arg odom_topic)"/>
        </node>
        <node pkg="tauv_common" type="position_controller" name="position_control"
              output="screen">
            <remap from="cmd_vel" to="controller/cmd_vel"/>
            <remap from="odom" to="$(arg odom_topic)"/>
        </node>

        <!-- Launch controller wrapper -->
        <node pkg="tauv_common" type="pid_control_wrapper" name="pid_control_wrapper" output="screen">
            <remap from="joy_cmd_acc" to="teleop/cmd_acc"/>
            <remap from="joy_cmd_vel" to="teleop/cmd_vel"/>
            <remap from="joy_cmd_pos" to="teleop/cmd_pos"/>
            <remap from="controller_cmd_acc" to="controller/cmd_acc"/>
            <remap from="controller_cmd_vel" to="controller/cmd_vel"/>
        </node>

        <!-- Load USB joystick controller configs -->
        <rosparam file="$(arg config_pkg)/joy.yaml" command="load"/>

        <!-- Start Teleop node -->
        <node pkg="tauv_common" type="teleop" name="teleop" output="screen">
            <remap from="cmd_acc" to="teleop/cmd_acc"/>
            <remap from="cmd_vel" to="teleop/cmd_vel"/>
            <remap from="cmd_pos" to="teleop/cmd_pos"/>
        </node>
    </group>



</launch>
